src/core/server/http/router/validator/validator.ts

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ValidationError, Type, schema, ObjectType } from '@kbn/config-schema';
+import { Stream } from 'stream';
+import { RouteValidationError } from './validator_error';
+
+/**
+ * Validation resolver to be used in the custom validation function
+ *
+ * See {@link RouteValidationFunction}.
+ *
+ * @public
+ */
+export interface RouteValidationResolver {
+  ok: <T>(value: T) => { value: T };
+  fail: (error: Error | string, path?: string[]) => { error: RouteValidationError };
+}
+
+/**
+ * The custom validation function if @kbn/config-schema is not a valid solution for your specific plugin requirements.
+ *
+ * The validation should look something like:
+ * ```typescript
+ * interface MyExpectedBody {
+ *   bar: string;
+ *   baz: number;
+ * }
+ *
+ * const myBodyValidation: RouteValidationFunction<MyExpectedBody> = (validationResolver, data) => {
+ *   const { ok, fail } = validationResolver;
+ *   const { bar, baz } = data || {};
+ *   if (typeof bar === 'string' && typeof baz === 'number') {
+ *     return ok({ bar, baz });
+ *   } else {
+ *     return fail('Wrong payload', ['body']);
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export type RouteValidationFunction<T> = (
+  validationResolver: RouteValidationResolver,
+  data: any
```

Comment:
I would expect these arguments to be in the reverse order to more closely match route handlers (where the input argument (request) comes before the response toolkit).

Reply:
I did it in that order to somehow enforce the user gets the param :)
But, in order to maintain our convention, I can switch them 👍